raft/log.go

```diff
@@ -1029,42 +1122,67 @@ func (l *Log) applier(done chan chan struct{}) {
 
 			// Verify, under lock, that we're not closing.
 			select {
-			case ch := <-done:
-				close(ch)
-				return errDone
+			case <-closing:
+				return nil
 			default:
 			}
 
 			// Ignore if there are no pending entries.
 			// Ignore if all entries are applied.
 			if len(l.entries) == 0 {
-				l.tracef("applier: no entries")
+				//l.tracef("applier: no entries")
 				return nil
 			} else if l.appliedIndex == l.commitIndex {
 				//l.tracef("applier: up to date")
 				return nil
 			}
 
-			// Calculate start index.
-			min := l.entries[0].Index
-			startIndex, endIndex := l.appliedIndex+1, l.commitIndex
-			if maxIndex := l.entries[len(l.entries)-1].Index; l.commitIndex > maxIndex {
-				l.tracef("applier: commit index above max: commit=%d, max=%d", l.commitIndex, maxIndex)
-				endIndex = maxIndex
+			// Determine the available range of indices on the log.
+			entmin, entmax := l.entries[0].Index, l.entries[len(l.entries)-1].Index
+			assert(entmin <= entmax, "apply: log out of order: min=%d, max=%d", entmin, entmax)
```

Comment:
Yes, these situations shouldn't occur. If they do then I'd rather it crash hard here rather than try to keep running and crash somewhere else or corrupt data.